workflow/scripts/transport/direct_damages.py

```diff
@@ -0,0 +1,213 @@
+
+    # calculate damages for assets we have damage curves for
+    damage_fraction_by_asset_type = []
+    logging.info(f"Exposed assets {set(exposure.asset_type)}")
+    for asset_type in set(exposure.asset_type) & set(damage_curves.keys()):
+
+        logging.info(f"Processing {asset_type=}")
+        damage_curve: pd.DataFrame = damage_curves[asset_type]
+
+        # pick out rows of asset type and columns of hazard intensity
+        asset_type_mask: gpd.GeoDataFrame = exposure.asset_type == asset_type
+        asset_exposure: pd.DataFrame = pd.DataFrame(exposure.loc[asset_type_mask, hazard_columns])
+
+        # create interpolated damage curve for given asset type
+        hazard_intensity, damage_fraction = damage_curve.iloc[:, 0], damage_curve.iloc[:, 1]
+        # if curve of length n, where x < x_0, y = y_0 and where x > x_n, y = y_n
+        bounds = tuple(f(damage_fraction) for f in (min, max))
+        interpolated_damage_curve = interp1d(hazard_intensity, damage_fraction, kind='linear', fill_value=bounds, bounds_error=False)
+
+        # apply damage_curve function to exposure table
+        # the return value of interpolated_damage_curve is a numpy array
+        logging.info("Calculating damage fractions")
+        damage_fraction_for_asset_type = pd.DataFrame(
+            interpolated_damage_curve(asset_exposure),
+            index=asset_exposure.index,
+            columns=asset_exposure.columns
+        )
+
+        # store the computed direct damages and any columns we started with
+        # (other than exposure)
+        damage_fraction_by_asset_type.append(
+            pd.concat(
+                [
+                    damage_fraction_for_asset_type,
+                    exposure.loc[asset_type_mask, non_hazard_columns]
+                ],
+                axis="columns"
+            )
+        )
+
+    # concatenate damage fractions for different asset types into single dataframe
+    damage_fraction: gpd.GeoDataFrame = gpd.GeoDataFrame(pd.concat(damage_fraction_by_asset_type))
+    # write to disk
+    logging.info(f"Writing {damage_fraction.shape=} to disk")
+    damage_fraction.to_parquet(damage_fraction_path)
+
+    # multiply the damage fraction estimates by a cost to rebuild the asset
+    # units are: 1 * USD/km * km = USD
+    logging.info("Calculating direct damage costs")
+    direct_damages_only = damage_fraction[hazard_columns] \
+            .multiply(damage_fraction[REHABILITATION_COST_FIELD], axis="index") \
+            .multiply(damage_fraction[SPLIT_LENGTH_FIELD], axis="index")
+
+    logging.info(f"Reading raw network data for unsplit geometry")
+    unsplit: gpd.GeoDataFrame = gpd.read_parquet(unsplit_path)
+    logging.info(f"{unsplit.shape=}")
+
+    # join the other fields with the direct damage estimates
+    logging.info("Unifying rasterised segments and summing damage costs")
+    direct_damages: gpd.GeoDataFrame = gpd.GeoDataFrame(
+        pd.concat(
```

Comment:
This concat feels a bit fragile, though I can convince myself it's right. What do you think about something like the below?

- stick an `edge_id` column back on `direct_damages_only` and `.groupby('edge_id').sum()`
- pick all `non_hazard_columns` (including `geometry`) from `unsplit` and set `edge_id` as index
- join the two

```python
grouped_direct_damages = pd.concat(
    [direct_damages_only, damage_fraction["edge_id"], 
    axis="columns"
).set_index("edge_id")
unsplit_subset = unsplit[non_hazard_columns].set_index("edge_id")
direct_damages = unsplit_subset.join(grouped_direct_damages, validate='one_to_one')
```

Reply:
There were _reasons_ why I did it this way, but yes I agree relying on the repeated index values seems more brittle than using the `edge_id`.

I think you're missing a call to groupby in the above block? I'll see if I can get something like your suggestion to work.